### ISSUE TITLE
How to get descriptions saved as comments?

### ISSUE THREAD

**Issue by Tobychev** [NowanIlfideme/pydantic-yaml#83]
In pydantic you can use the Field function to attach a description to the model properties, and I wonder if there is a way that pydantic-yaml could save that description as a comment?

**Comment by NowanIlfideme** [comment]
Hi, currently that isn't possible through `pydantic-yaml`. It could be possible to do through the YAML backend we use, `ruamel.yaml` (see [here](https://yaml.readthedocs.io/en/latest/detail.html#adding-replacing-comments) for details), but currently the loading/dumping is done through JSON first, which doesn't support comments.

This is functionality I want as well, but so far the only way I've figured out would be writing a custom YAML engine (possibly on top of `ruamel.yaml`), which is... somewhere in the pipeline, but not now. 😅

**Comment by NowanIlfideme** [comment]
Well, yes, you can look through the `ModelField` classes and add comments. However, you need to be able to map the model fields to the JSON Path in the resulting mess of dict/list/scalars from the JSON.

There's also the consideration of how you'd annotate it...
Do you apply the comment to only the field value? If it's a list, where do you anchor the documentation? Should you output description as a comment for all fields, or just a subset? I have some opinions on this, but for usability it should really be configurable.

If you have a simple use-case right now (i.e. just one or more nested Pydantic models with fields), it is actually not that difficult to implement - you can use `model.dict()` (or the equivalent in Pydantic v2, not sure - `model_dict()`?), pass it to a ruamel.yaml `CommentedMap`, then add comments from the Pydantic fields. Or - even better - you can write it with `pydantic-yaml` to a temporary buffer, read with `ruamel.yaml` with a round-trip-capable serializer, and add comments through that interface.

Making this work _in the general case_ is an issue, since you can have containers (lists, dicts, list-like, dict-like, ...) of Pydantic and non-Pydantic objects. And all can have field validations.

I hope I will find something in `pydantic-core` that will help with this, once I get to it. I'll be on vacation soon, so mayyyyybe...

**Comment by ipcoder** [comment]
Up-voting this! 
:-)

**Comment by scottstanie** [comment]
Hi, I believe I started looking into this before this project got going: https://github.com/opera-adt/dolphin/blob/main/src/dolphin/workflows/_yaml_model.py
I created the `YamlModel` for my project which has `to_yaml` and `from_yaml` to load and save.

Since that part is already here, perhaps someone would be able to take the `_add_comments` function from here https://github.com/opera-adt/dolphin/blob/f9f2f87e83140f43550c0842aa140ac9f01cab43/src/dolphin/workflows/_yaml_model.py#L136

This for Pydantic V1, and i'm working on updating things (if there were any breaking changes) to V2 shortly.

**Comment by richmr** [comment]
Hi there.  Came across this discussion after having the same functional goal of writing my models and having them make self-documenting YAML templates.  I spent some time looking at the pydantic-yaml code and the underlying ruamel.yaml library.  I'm sure a plugin or subclass could make this happen, but it looked like a lot to learn and sort of reverse-engineer.

Instead I [wrote this gist](https://gist.github.com/richmr/5762736da29fd33f544f3eaf4dcd1f7f) that works at the file level and leverages the original models used to make the template file.

I'm sure there are a TON of use cases the gist doesn't cover but its working for my needs.

Hope it helps!

**Comment by fer-rum** [comment]
Indeed, Adding comments to the YAML file from a fields' description (or docstring) would be amazing.
The approach of @richmr already looks quite promising.

**Comment by Shua-github** [comment]
```python
class ModelToCommentedMap:
    def __init__(
        self,
        mapping_indent: int = 2,
        custom_handlers: Optional[dict[type, Callable[[Any], Any]]] = None,  # noqa: FA100
    ) -> None:
        self.mapping_indent = mapping_indent
        self.custom_handlers = custom_handlers or {}

    def register_handler(self, type_: type, handler: Callable[[Any], Any]) -> None:
        self.custom_handlers[type_] = handler

    def convert(self, model: BaseModel) -> CommentedMap:
        return self._convert_model(model, _indent=0)

    def _convert_model(self, model: BaseModel, _indent: int) -> CommentedMap:
        commented = CommentedMap()
        fields = type(model).model_fields

        for name, field in fields.items():
            value = getattr(model, name)
            commented[name] = self._convert_value(value, _indent + self.mapping_indent)

            if field.description:
                commented.yaml_set_comment_before_after_key(name, before=field.description, indent=_indent)

        return commented

    def _convert_value(self, value: Any, _indent: int) -> Any:  # noqa: ANN401
        for type_, handler in self.custom_handlers.items():
            if isinstance(value, type_):
                return handler(value)

        if isinstance(value, BaseModel):
            return self._convert_model(value, _indent)

        if isinstance(value, dict):
            nested_map = CommentedMap()
            for k, v in value.items():
                nested_map[k] = self._convert_value(v, _indent + self.mapping_indent)
            return nested_map

        if isinstance(value, Path):
            return str(value)

        return value
```
I have written a class that is currently sufficient for my use (hope this helps with the issue)

**Comment by NowanIlfideme** [comment]
Hi all, post-processing the file with comments (like in [this gist](https://gist.github.com/richmr/5762736da29fd33f544f3eaf4dcd1f7f) linked above) is one way, though the issue is always going to be with nested models (`foo: MyModel`) and sub-fields (like `list[MyModel]` or `dict[str, MyModel]`), and even moreso with optional or union fields (`foo: MyModel1 | MyModel2 | None = None`). Which is why it's been so long & the feature isn't in `pydantic-yaml` yet.

Thankfully, writing a YAML dumper is easier than a YAML parser, but it still takes quite a bit of time. I might find some time on my hands to implement this, as I also would use this feature a lot. :)

**Comment by NowanIlfideme** [comment]
I also found this project: [pydantic-examples](https://pypi.org/project/pydantic-examples/), specifically [this file](https://github.com/bneijt/pydantic-examples/blob/main/src/pydantic_examples/yaml.py) based on ruamel.yaml and adding comments by post-processing.

I've reached out [here](https://github.com/bneijt/pydantic-examples/issues/10) about inclusion.

**Comment by NowanIlfideme** [comment]
Hi all, please check out the [latest alpha release 1.6.0a1](https://github.com/NowanIlfideme/pydantic-yaml/releases/tag/v1.6.0-alpha.1) (also [on PyPI](https://pypi.org/project/pydantic-yaml/1.6.0a1/)). You can try by pinning `pydantic-yaml~=1.6.0a1`.

There are not enough unit tests, and frankly the `ruamel.yaml` comment-adding implementation is... hacky. But it works about as well as possible without the custom YAML engine... 😉
